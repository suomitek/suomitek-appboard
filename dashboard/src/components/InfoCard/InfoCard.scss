.ListItem {
  color: #1c2b39;
}

.ListItem__content__title {
  margin: 0;
  width: 100%;
  font-weight: bold;
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: nowrap;
  color: #1c2b39;
}

.ListItem__content__info {
  display: flex;
  justify-content: space-between;
}

.ListItem__content__info_tag {
  float: right;
  margin-left: 3px;
  margin-bottom: 3px;
  color: #fff;
  border-radius: 16px;
  max-width: 120px;
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: nowrap;
}

.ListItem__content__info_tag-2 {
  background-color: #1c2b39;
}

.ListItem__content__info_tag-1 {
  background-color: #1598cb;
}

.ListItem__header {
  position: relative;
  display: block;
}

.ListItem__banner {
  position: absolute;
  bottom: 0;
  background: #008145;
  color: white;
  width: 100%;
  left: 0;
  margin: 0;
  text-align: center;
  padding: 0.25em 0;
  font-size: 14px;
}
