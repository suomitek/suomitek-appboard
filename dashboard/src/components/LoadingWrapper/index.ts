import LoadingWrapper, { LoaderType } from "./LoadingWrapper";
export { LoaderType };
export default LoadingWrapper;
