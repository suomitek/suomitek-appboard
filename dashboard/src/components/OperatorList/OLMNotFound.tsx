import * as React from "react";

import { NotFoundErrorAlert } from "../ErrorAlert";

class OLMNotFound extends React.Component {
  public render() {
    return (
      <NotFoundErrorAlert header="Operator Lifecycle Manager (OLM) not installed.">
        <div>
          <p>
            Ask an administrator to install the{" "}
            <a
              href="https://github.com/operator-framework/operator-lifecycle-manager"
              target="_blank"
              rel="noopener noreferrer"
            >
              Operator Lifecycle Manager
            </a>{" "}
            to browse, provision and manage Operators within Suomitek-appboard.
          </p>
          To install the OLM, execute the following command in a terminal with <code>kubectl</code>{" "}
          available and configured:
          <section className="AppNotes Terminal elevation-1 margin-v-big">
            <div className="Terminal__Top type-small">
              <div className="Terminal__Top__Buttons">
                <span className="Terminal__Top__Button Terminal__Top__Button--red" />
                <span className="Terminal__Top__Button Terminal__Top__Button--yellow" />
                <span className="Terminal__Top__Button Terminal__Top__Button--green" />
              </div>
            </div>
            <div className="Terminal__Tab">
              <pre className="Terminal__Code">
                <code>
                  curl -L
                  https://github.com/operator-framework/operator-lifecycle-manager/releases/download/0.15.1/install.sh
                  -o install.sh <br />
                  chmod +x install.sh <br />
                  ./install.sh 0.15.1
                </code>
              </pre>
            </div>
          </section>
        </div>
      </NotFoundErrorAlert>
    );
  }
}

export default OLMNotFound;
