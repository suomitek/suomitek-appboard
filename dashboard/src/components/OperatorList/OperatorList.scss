.horizontal-column {
  border-right: 1px solid lightgray;
}
