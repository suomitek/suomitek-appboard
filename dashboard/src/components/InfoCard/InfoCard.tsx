import * as React from "react";

import { Link } from "react-router-dom";
import placeholder from "../../placeholder.png";
import Card, { CardContent, CardIcon } from "../Card";
import "./InfoCard.css";

export interface IServiceInstanceCardProps {
  title: string;
  info: string;
  link?: string;
  icon?: string;
  banner?: string;
  subIcon?: string;
  description?: string | JSX.Element;
  tag1Class?: string;
  tag1Content?: string | JSX.Element;
  tag2Class?: string;
  tag2Content?: string | JSX.Element;
}

const InfoCard: React.SFC<IServiceInstanceCardProps> = props => {
  const {
    title,
    link,
    info,
    description,
    tag1Content,
    tag1Class,
    tag2Content,
    tag2Class,
    banner,
    subIcon,
  } = props;
  const icon = props.icon ? props.icon : placeholder;
  return (
    <Card responsive={true} className="ListItem">
      <Link to={link || "#"} title={title} className="ListItem__header">
        <CardIcon icon={icon} />
        {banner && <p className="ListItem__banner">{banner}</p>}
        {subIcon && (
          <div className="Card__subIcon bg-light text-r">
            <img src={subIcon} alt="icon" />
          </div>
        )}
      </Link>
      <CardContent>
        <div className="ListItem__content">
          <Link to={link || "#"} title={title}>
            <h3 className="ListItem__content__title type-big">{title}</h3>
          </Link>
          {description}
          <div className="ListItem__content__info">
            <p className="margin-reset type-small padding-t-tiny type-color-light-blue">{info}</p>
            <div>
              {tag1Content && (
                <span
                  className={`ListItem__content__info_tag ListItem__content__info_tag-1 type-small type-color-white padding-t-tiny padding-h-normal ${tag1Class ||
                    ""}`}
                >
                  {tag1Content}
                </span>
              )}
              {tag2Content && (
                <span
                  className={`ListItem__content__info_tag ListItem__content__info_tag-2 type-small type-color-white padding-t-tiny padding-h-normal margin-b-normal ${tag2Class ||
                    ""}`}
                >
                  {tag2Content}
                </span>
              )}
            </div>
          </div>
        </div>
      </CardContent>
      {props.children}
    </Card>
  );
};

export default InfoCard;
