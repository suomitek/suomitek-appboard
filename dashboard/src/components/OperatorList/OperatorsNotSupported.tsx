import { MessageAlert } from "components/ErrorAlert";
import * as React from "react";
import { Link } from "react-router-dom";

import * as url from "shared/url";

interface IOperatorNotSupportedProps {
  namespace: string;
}

function OperatorNotSupported(props: IOperatorNotSupportedProps) {
  return (
    <MessageAlert header="Operators are supported on the default cluster only">
      <div>
        <p className="margin-v-normal">
          Kubeapps' Operator support enables the addition of{" "}
          <Link to={url.app.operators.list("default", props.namespace)}>
            operators on the default cluster only
          </Link>
          .
        </p>
      </div>
    </MessageAlert>
  );
}

export default OperatorNotSupported;
