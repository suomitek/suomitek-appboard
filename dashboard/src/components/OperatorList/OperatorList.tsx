import { RouterAction } from "connected-react-router";
import { flatten, intersection, uniq } from "lodash";
import * as React from "react";

import {
  ForbiddenError,
  IClusterServiceVersion,
  IPackageManifest,
  IPackageManifestStatus,
} from "../../shared/types";
import { api, app } from "../../shared/url";
import { CardGrid } from "../Card";
import { ErrorSelector, MessageAlert } from "../ErrorAlert";
import InfoCard from "../InfoCard";
import LoadingWrapper from "../LoadingWrapper";
import {
  AUTO_PILOT,
  BASIC_INSTALL,
  DEEP_INSIGHTS,
  FULL_LIFECYCLE,
  SEAMLESS_UPGRADES,
} from "../OperatorView/OperatorCapabilityLevel";
import PageHeader from "../PageHeader";
import SearchFilter from "../SearchFilter";
import OLMNotFound from "./OLMNotFound";
import OperatorNotSupported from "./OperatorsNotSupported";

import "./OperatorList.css";

export interface IOperatorListProps {
  isFetching: boolean;
  checkOLMInstalled: (namespace: string) => Promise<boolean>;
  isOLMInstalled: boolean;
  cluster: string;
  namespace: string;
  getOperators: (namespace: string) => Promise<void>;
  operators: IPackageManifest[];
  error?: Error;
  getCSVs: (namespace: string) => Promise<IClusterServiceVersion[]>;
  csvs: IClusterServiceVersion[];
  filter: string;
  pushSearchFilter: (filter: string) => RouterAction;
}

export interface IOperatorListState {
  filter: string;
  categories: string[];
  filterCategories: { [key: string]: boolean };
  filterCapabilities: { [key: string]: boolean };
}

function getDefaultChannel(packageStatus: IPackageManifestStatus) {
  const defaultChannel = packageStatus.defaultChannel;
  const channel = packageStatus.channels.find(ch => ch.name === defaultChannel);
  return channel!;
}

function getCategories(packageStatus: IPackageManifestStatus) {
  const channel = getDefaultChannel(packageStatus);
  return channel.currentCSVDesc.annotations.categories.split(",").map(c => c.trim());
}

function getCapabilities(packageStatus: IPackageManifestStatus) {
  const channel = getDefaultChannel(packageStatus);
  return channel.currentCSVDesc.annotations.capabilities;
}

class OperatorList extends React.Component<IOperatorListProps, IOperatorListState> {
  public state: IOperatorListState = {
    filter: "",
    categories: [],
    filterCategories: {},
    filterCapabilities: {
      [BASIC_INSTALL]: false,
      [SEAMLESS_UPGRADES]: false,
      [FULL_LIFECYCLE]: false,
      [DEEP_INSIGHTS]: false,
      [AUTO_PILOT]: false,
    },
  };

  public componentDidMount() {
    this.props.checkOLMInstalled(this.props.namespace);
    this.props.getOperators(this.props.namespace);
    this.props.getCSVs(this.props.namespace);
    this.setState({ filter: this.props.filter });
  }

  public componentDidUpdate(prevProps: IOperatorListProps) {
    if (prevProps.namespace !== this.props.namespace) {
      this.props.getOperators(this.props.namespace);
      this.props.getCSVs(this.props.namespace);
    }
    if (this.props.filter !== prevProps.filter) {
      this.props.getOperators(this.props.namespace);
      this.props.getCSVs(this.props.namespace);
      this.setState({ filter: this.props.filter });
    }

    if (this.props.operators !== prevProps.operators) {
      const categories = uniq(
        flatten(this.props.operators.map(operator => getCategories(operator.status))),
      );
      const filterCategories = {};
      categories.forEach(category => {
        filterCategories[category] = false;
      });
      this.setState({ categories, filterCategories });
    }
  }

  public render() {
    const { cluster, namespace, isFetching, pushSearchFilter } = this.props;
    if (cluster !== "default") {
      return <OperatorNotSupported namespace={namespace} />;
    }
    return (
      <div>
        <PageHeader>
          <h1>Operators</h1>
          <SearchFilter
            className="margin-l-big"
            placeholder="search operators..."
            onChange={this.handleFilterQueryChange}
            value={this.state.filter}
            onSubmit={pushSearchFilter}
          />
        </PageHeader>
        <main>
          <MessageAlert level="warning">
            <div>
              Operators integration is under heavy development and currently in alpha state. If you
              find an issue please report it{" "}
              <a
                target="_blank"
                rel="noopener noreferrer"
                href="https://github.com/kubeapps/kubeapps/issues"
              >
                here.
              </a>
            </div>
          </MessageAlert>
          <LoadingWrapper loaded={!isFetching}>{this.renderOperators()}</LoadingWrapper>
        </main>
      </div>
    );
  }

  private renderOperators() {
    const { operators, error, csvs, isOLMInstalled } = this.props;
    const { filter, filterCategories, filterCapabilities } = this.state;
    if (error && error.constructor === ForbiddenError) {
      return (
        <ErrorSelector
          error={error}
          action="list"
          resource="Operators"
          namespace={this.props.namespace}
        />
      );
    }
    if (!isOLMInstalled) {
      return <OLMNotFound />;
    }
    if (error) {
      return (
        <ErrorSelector
          error={error}
          action="list"
          resource="Operators"
          namespace={this.props.namespace}
        />
      );
    }
    const csvNames = csvs.map(csv => csv.metadata.name);
    const installedOperators: IPackageManifest[] = [];
    const availableOperators: IPackageManifest[] = [];
    const filteredOperators = operators.filter(operator => {
      if (filter && !operator.metadata.name.match(filter)) {
        return false;
      }
      const hasFilteredCategories = Object.values(filterCategories).some(
        filteredCategory => filteredCategory,
      );
      if (hasFilteredCategories) {
        const allowedCategories = Object.keys(filterCategories).filter(
          cat => filterCategories[cat],
        );
        const categories = getCategories(operator.status);
        if (intersection(allowedCategories, categories).length === 0) {
          return false;
        }
      }
      const hasFilteredCapabilities = Object.values(filterCapabilities).some(
        filterCapability => filterCapability,
      );
      if (hasFilteredCapabilities) {
        const allowedCapabilities = Object.keys(filterCapabilities).filter(
          capability => filterCapabilities[capability],
        );
        if (
          !allowedCapabilities.some(capability => capability === getCapabilities(operator.status))
        ) {
          return false;
        }
      }
      return true;
    });
    filteredOperators.forEach(operator => {
      if (csvNames.some(csvName => csvName === getDefaultChannel(operator.status).currentCSV)) {
        installedOperators.push(operator);
      } else {
        availableOperators.push(operator);
      }
    });
    return (
      <div className="row margin-t-big">
        <div className="col-2 margin-t-big horizontal-column">
          <div className="margin-b-normal ">
            <b>Categories</b>
          </div>
          {this.state.categories.map(category => {
            return (
              <div key={category}>
                <label
                  className="checkbox"
                  key={category}
                  onChange={this.toggleFilterCategory(category)}
                >
                  <input type="checkbox" />
                  <span>{category}</span>
                </label>
              </div>
            );
          })}
          <div className="margin-v-normal ">
            <b>Capability Level</b>
          </div>
          {[BASIC_INSTALL, SEAMLESS_UPGRADES, FULL_LIFECYCLE, DEEP_INSIGHTS, AUTO_PILOT].map(
            capability => {
              return (
                <div key={capability}>
                  <label
                    className="checkbox"
                    key={capability}
                    onChange={this.toggleFilterCapability(capability)}
                  >
                    <input type="checkbox" />
                    <span>{capability}</span>
                  </label>
                </div>
              );
            },
          )}
        </div>
        <div className="col-10">
          <div className="padding-l-normal">
            {filteredOperators.length === 0 ? (
              <p>No Operator found</p>
            ) : (
              this.renderCardGrid(installedOperators, availableOperators)
            )}
          </div>
        </div>
      </div>
    );
  }

  private renderCardGrid(
    installedOperators: IPackageManifest[],
    availableOperators: IPackageManifest[],
  ) {
    return (
      <>
        {installedOperators.length > 0 && (
          <>
            <h3>Installed</h3>
            <CardGrid>
              {installedOperators.map(operator => {
                return (
                  <InfoCard
                    key={operator.metadata.name}
                    link={app.operators.view(
                      this.props.cluster,
                      this.props.namespace,
                      operator.metadata.name,
                    )}
                    title={operator.metadata.name}
                    icon={api.operators.operatorIcon(this.props.namespace, operator.metadata.name)}
                    info={`v${operator.status.channels[0].currentCSVDesc.version}`}
                    tag1Content={operator.status.channels[0].currentCSVDesc.annotations.categories}
                    tag2Content={operator.status.provider.name}
                  />
                );
              })}
            </CardGrid>
          </>
        )}
        <h3>Available Operators</h3>
        <CardGrid>
          {availableOperators.map(operator => {
            return (
              <InfoCard
                key={operator.metadata.name}
                link={app.operators.view(
                  this.props.cluster,
                  this.props.namespace,
                  operator.metadata.name,
                )}
                title={operator.metadata.name}
                icon={api.operators.operatorIcon(this.props.namespace, operator.metadata.name)}
                info={`v${operator.status.channels[0].currentCSVDesc.version}`}
                tag1Content={operator.status.channels[0].currentCSVDesc.annotations.categories}
                tag2Content={operator.status.provider.name}
              />
            );
          })}
        </CardGrid>
      </>
    );
  }

  private handleFilterQueryChange = (filter: string) => {
    this.setState({
      filter,
    });
  };

  private toggleFilterCategory = (category: string) => {
    return () => {
      const { filterCategories } = this.state;
      this.setState({
        filterCategories: {
          ...filterCategories,
          [category]: !filterCategories[category],
        },
      });
    };
  };

  private toggleFilterCapability = (capability: string) => {
    return () => {
      const { filterCapabilities } = this.state;
      this.setState({
        filterCapabilities: {
          ...filterCapabilities,
          [capability]: !filterCapabilities[capability],
        },
      });
    };
  };
}

export default OperatorList;
