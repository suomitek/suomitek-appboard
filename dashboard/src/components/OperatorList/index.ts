import OperatorList from "./OperatorList";

export default OperatorList;
