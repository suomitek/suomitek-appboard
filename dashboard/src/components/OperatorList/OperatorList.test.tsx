import { shallow } from "enzyme";
import * as React from "react";
import itBehavesLike from "../../shared/specs";
import { ForbiddenError, IPackageManifest } from "../../shared/types";
import { CardGrid } from "../Card";
import { ErrorSelector } from "../ErrorAlert";
import InfoCard from "../InfoCard";
import LoadingWrapper from "../LoadingWrapper";
import { AUTO_PILOT, BASIC_INSTALL } from "../OperatorView/OperatorCapabilityLevel";
import OLMNotFound from "./OLMNotFound";
import OperatorList, { IOperatorListProps } from "./OperatorList";
import OperatorNotSupported from "./OperatorsNotSupported";

const defaultProps: IOperatorListProps = {
  isFetching: false,
  checkOLMInstalled: jest.fn(),
  isOLMInstalled: false,
  operators: [],
  cluster: "default",
  namespace: "default",
  getOperators: jest.fn(),
  getCSVs: jest.fn(),
  csvs: [],
  filter: "",
  pushSearchFilter: jest.fn(),
};

const sampleOperator = {
  metadata: {
    name: "foo",
  },
  status: {
    provider: {
      name: "kubeapps",
    },
    defaultChannel: "alpha",
    channels: [
      {
        name: "alpha",
        currentCSV: "kubeapps-operator",
        currentCSVDesc: {
          version: "1.0.0",
          annotations: {
            categories: "security",
            capabilities: AUTO_PILOT,
          },
        },
      },
    ],
  },
} as IPackageManifest;

const sampleCSV = {
  metadata: { name: "kubeapps-operator" },
  spec: {
    icon: [{}],
    provider: {
      name: "kubeapps",
    },
    customresourcedefinitions: {
      owned: [
        {
          name: "foo.kubeapps.com",
          version: "v1alpha1",
          kind: "Foo",
          resources: [{ kind: "Deployment" }],
        },
      ],
    },
  },
} as any;

itBehavesLike("aLoadingComponent", {
  component: OperatorList,
  props: { ...defaultProps, isFetching: true },
});

it("call the OLM check and render the NotFound message if not found", () => {
  const checkOLMInstalled = jest.fn();
  const wrapper = shallow(<OperatorList {...defaultProps} checkOLMInstalled={checkOLMInstalled} />);
  expect(checkOLMInstalled).toHaveBeenCalled();
  expect(wrapper.find(OLMNotFound)).toExist();
});

it("displays an alert if rendered for an additional cluster", () => {
  const props = { ...defaultProps, cluster: "other-cluster" };
  const wrapper = shallow(<OperatorList {...props} />);
  expect(wrapper.find(OperatorNotSupported)).toExist();
});

it("call the OLM check and render the a Forbidden message if it could do the check", () => {
  const checkOLMInstalled = jest.fn();
  const wrapper = shallow(<OperatorList {...defaultProps} checkOLMInstalled={checkOLMInstalled} />);
  wrapper.setProps({ error: new ForbiddenError("nope") });
  expect(checkOLMInstalled).toHaveBeenCalled();
  expect(wrapper.find(ErrorSelector)).toExist();
  expect(wrapper.find(OLMNotFound)).not.toExist();
});

it("re-request operators if the namespace changes", () => {
  const getOperators = jest.fn();
  const getCSVs = jest.fn();
  const wrapper = shallow(
    <OperatorList {...defaultProps} getOperators={getOperators} getCSVs={getCSVs} />,
  );
  wrapper.setProps({ namespace: "other" });
  expect(getOperators).toHaveBeenCalledTimes(2);
  expect(getCSVs).toHaveBeenCalledTimes(2);
});

it("renders an error if exists", () => {
  const wrapper = shallow(
    <OperatorList {...defaultProps} isOLMInstalled={true} error={new Error("Boom!")} />,
  );
  expect(wrapper.find(ErrorSelector)).toExist();
  expect(
    wrapper
      .find(ErrorSelector)
      .dive()
      .dive()
      .text(),
  ).toMatch("Boom!");
});

it("skips the error if the OLM is not installed", () => {
  const wrapper = shallow(
    <OperatorList
      {...defaultProps}
      isOLMInstalled={false}
      error={new Error("There are no operators!")}
    />,
  );
  expect(wrapper.find(ErrorSelector)).not.toExist();
  expect(wrapper.find(OLMNotFound)).toExist();
});

it("render the operator list with installed operators", () => {
  const wrapper = shallow(
    <OperatorList
      {...defaultProps}
      isOLMInstalled={true}
      operators={[sampleOperator]}
      csvs={[sampleCSV]}
    />,
  );
  expect(wrapper.find(OLMNotFound)).not.toExist();
  expect(wrapper.find(InfoCard)).toExist();
  // The section "Available operators" should be empty since all the ops are installed
  expect(wrapper.find("h3").filterWhere(c => c.text() === "Installed")).toExist();
  expect(
    wrapper
      .find(CardGrid)
      .last()
      .children(),
  ).not.toExist();
  expect(wrapper).toMatchSnapshot();
});

it("render the operator list without installed operators", () => {
  const wrapper = shallow(
    <OperatorList {...defaultProps} isOLMInstalled={true} operators={[sampleOperator]} csvs={[]} />,
  );
  expect(wrapper.find(OLMNotFound)).not.toExist();
  expect(wrapper.find(InfoCard)).toExist();
  // The section "Available operators" should not be empty since the operator is not installed
  expect(wrapper.find("h3").filterWhere(c => c.text() === "Installed")).not.toExist();
  expect(
    wrapper
      .find(CardGrid)
      .last()
      .children(),
  ).toExist();
  expect(wrapper).toMatchSnapshot();
});

describe("filter operators", () => {
  const sampleOperator2 = {
    ...sampleOperator,
    metadata: {
      name: "bar",
    },
    status: {
      ...sampleOperator.status,
      channels: [
        {
          ...sampleOperator.status.channels[0],
          currentCSVDesc: {
            version: "1.0.0",
            annotations: {
              categories: "database, other",
              capabilities: BASIC_INSTALL,
            },
          },
        },
      ],
    },
  } as any;

  it("setting the filter in the state", () => {
    const wrapper = shallow(
      <OperatorList
        {...defaultProps}
        isOLMInstalled={true}
        operators={[sampleOperator, sampleOperator2]}
        csvs={[]}
      />,
    );
    expect(wrapper.find(InfoCard).length).toBe(2);
    wrapper.setState({ filter: "foo" });
    expect(wrapper.find(InfoCard).length).toBe(1);
  });

  it("setting the filter in the props", () => {
    const wrapper = shallow(
      <OperatorList
        {...defaultProps}
        isOLMInstalled={true}
        operators={[sampleOperator, sampleOperator2]}
        csvs={[]}
      />,
    );
    expect(wrapper.find(InfoCard).length).toBe(2);
    wrapper.setProps({ filter: "foo" });
    expect(wrapper.find(InfoCard).length).toBe(1);
  });

  it("show a message if the filter doesn't match any operator", () => {
    const wrapper = shallow(
      <OperatorList
        {...defaultProps}
        isOLMInstalled={true}
        operators={[sampleOperator, sampleOperator2]}
        csvs={[]}
      />,
    );
    expect(wrapper.find(InfoCard).length).toBe(2);
    wrapper.setProps({ filter: "nope" });
    expect(wrapper.find(InfoCard)).not.toExist();
    expect(
      wrapper
        .find(LoadingWrapper)
        .dive()
        .text(),
    ).toMatch("No Operator found");
    expect(wrapper.find(".horizontal-column")).toExist();
  });

  it("filters by category", () => {
    const wrapper = shallow(<OperatorList {...defaultProps} isOLMInstalled={true} csvs={[]} />);
    wrapper.setProps({ operators: [sampleOperator, sampleOperator2] });
    const column = wrapper.find(".horizontal-column").text();
    expect(column).toContain("security");
    expect(column).toContain("database");
    expect(column).toContain("other");
    expect(wrapper.find(InfoCard).length).toBe(2);

    // Filter category "security"
    wrapper.setState({
      filterCategories: {
        security: true,
      },
    });
    expect(wrapper.find(InfoCard).length).toBe(1);
  });

  it("filters by capability", () => {
    const wrapper = shallow(<OperatorList {...defaultProps} isOLMInstalled={true} csvs={[]} />);
    wrapper.setProps({ operators: [sampleOperator, sampleOperator2] });
    expect(wrapper.find(InfoCard).length).toBe(2);

    // Filter by capability "Basic Install"
    wrapper.setState({
      filterCapabilities: {
        [BASIC_INSTALL]: true,
      },
    });
    expect(wrapper.find(InfoCard).length).toBe(1);
  });
});
