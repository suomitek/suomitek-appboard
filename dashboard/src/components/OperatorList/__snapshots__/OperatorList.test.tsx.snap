// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loading spinner matches the snapshot 1`] = `
<div>
  <PageHeader>
    <h1>
      Operators
    </h1>
    <SearchFilter
      className="margin-l-big"
      onChange={[Function]}
      onSubmit={[MockFunction]}
      placeholder="search operators..."
      value=""
    />
  </PageHeader>
  <main>
    <MessageAlertPage
      level="warning"
    >
      <div>
        Operators integration is under heavy development and currently in alpha state. If you find an issue please report it
         
        <a
          href="https://github.com/kubeapps/kubeapps/issues"
          rel="noopener noreferrer"
          target="_blank"
        >
          here.
        </a>
      </div>
    </MessageAlertPage>
    <LoadingWrapper
      loaded={false}
      type={0}
    >
      <OLMNotFound />
    </LoadingWrapper>
  </main>
</div>
`;

exports[`render the operator list with installed operators 1`] = `
<div>
  <PageHeader>
    <h1>
      Operators
    </h1>
    <SearchFilter
      className="margin-l-big"
      onChange={[Function]}
      onSubmit={[MockFunction]}
      placeholder="search operators..."
      value=""
    />
  </PageHeader>
  <main>
    <MessageAlertPage
      level="warning"
    >
      <div>
        Operators integration is under heavy development and currently in alpha state. If you find an issue please report it
         
        <a
          href="https://github.com/kubeapps/kubeapps/issues"
          rel="noopener noreferrer"
          target="_blank"
        >
          here.
        </a>
      </div>
    </MessageAlertPage>
    <LoadingWrapper
      loaded={true}
      type={0}
    >
      <div
        className="row margin-t-big"
      >
        <div
          className="col-2 margin-t-big horizontal-column"
        >
          <div
            className="margin-b-normal "
          >
            <b>
              Categories
            </b>
          </div>
          <div
            className="margin-v-normal "
          >
            <b>
              Capability Level
            </b>
          </div>
          <div
            key="Basic Install"
          >
            <label
              className="checkbox"
              key="Basic Install"
              onChange={[Function]}
            >
              <input
                type="checkbox"
              />
              <span>
                Basic Install
              </span>
            </label>
          </div>
          <div
            key="Seamless Upgrades"
          >
            <label
              className="checkbox"
              key="Seamless Upgrades"
              onChange={[Function]}
            >
              <input
                type="checkbox"
              />
              <span>
                Seamless Upgrades
              </span>
            </label>
          </div>
          <div
            key="Full Lifecycle"
          >
            <label
              className="checkbox"
              key="Full Lifecycle"
              onChange={[Function]}
            >
              <input
                type="checkbox"
              />
              <span>
                Full Lifecycle
              </span>
            </label>
          </div>
          <div
            key="Deep Insights"
          >
            <label
              className="checkbox"
              key="Deep Insights"
              onChange={[Function]}
            >
              <input
                type="checkbox"
              />
              <span>
                Deep Insights
              </span>
            </label>
          </div>
          <div
            key="Auto Pilot"
          >
            <label
              className="checkbox"
              key="Auto Pilot"
              onChange={[Function]}
            >
              <input
                type="checkbox"
              />
              <span>
                Auto Pilot
              </span>
            </label>
          </div>
        </div>
        <div
          className="col-10"
        >
          <div
            className="padding-l-normal"
          >
            <h3>
              Installed
            </h3>
            <CardGrid>
              <InfoCard
                icon="api/v1/namespaces/default/operator/foo/logo"
                info="v1.0.0"
                key="foo"
                link="/c/default/ns/default/operators/foo"
                tag1Content="security"
                tag2Content="kubeapps"
                title="foo"
              />
            </CardGrid>
            <h3>
              Available Operators
            </h3>
            <CardGrid />
          </div>
        </div>
      </div>
    </LoadingWrapper>
  </main>
</div>
`;

exports[`render the operator list without installed operators 1`] = `
<div>
  <PageHeader>
    <h1>
      Operators
    </h1>
    <SearchFilter
      className="margin-l-big"
      onChange={[Function]}
      onSubmit={[MockFunction]}
      placeholder="search operators..."
      value=""
    />
  </PageHeader>
  <main>
    <MessageAlertPage
      level="warning"
    >
      <div>
        Operators integration is under heavy development and currently in alpha state. If you find an issue please report it
         
        <a
          href="https://github.com/kubeapps/kubeapps/issues"
          rel="noopener noreferrer"
          target="_blank"
        >
          here.
        </a>
      </div>
    </MessageAlertPage>
    <LoadingWrapper
      loaded={true}
      type={0}
    >
      <div
        className="row margin-t-big"
      >
        <div
          className="col-2 margin-t-big horizontal-column"
        >
          <div
            className="margin-b-normal "
          >
            <b>
              Categories
            </b>
          </div>
          <div
            className="margin-v-normal "
          >
            <b>
              Capability Level
            </b>
          </div>
          <div
            key="Basic Install"
          >
            <label
              className="checkbox"
              key="Basic Install"
              onChange={[Function]}
            >
              <input
                type="checkbox"
              />
              <span>
                Basic Install
              </span>
            </label>
          </div>
          <div
            key="Seamless Upgrades"
          >
            <label
              className="checkbox"
              key="Seamless Upgrades"
              onChange={[Function]}
            >
              <input
                type="checkbox"
              />
              <span>
                Seamless Upgrades
              </span>
            </label>
          </div>
          <div
            key="Full Lifecycle"
          >
            <label
              className="checkbox"
              key="Full Lifecycle"
              onChange={[Function]}
            >
              <input
                type="checkbox"
              />
              <span>
                Full Lifecycle
              </span>
            </label>
          </div>
          <div
            key="Deep Insights"
          >
            <label
              className="checkbox"
              key="Deep Insights"
              onChange={[Function]}
            >
              <input
                type="checkbox"
              />
              <span>
                Deep Insights
              </span>
            </label>
          </div>
          <div
            key="Auto Pilot"
          >
            <label
              className="checkbox"
              key="Auto Pilot"
              onChange={[Function]}
            >
              <input
                type="checkbox"
              />
              <span>
                Auto Pilot
              </span>
            </label>
          </div>
        </div>
        <div
          className="col-10"
        >
          <div
            className="padding-l-normal"
          >
            <h3>
              Available Operators
            </h3>
            <CardGrid>
              <InfoCard
                icon="api/v1/namespaces/default/operator/foo/logo"
                info="v1.0.0"
                key="foo"
                link="/c/default/ns/default/operators/foo"
                tag1Content="security"
                tag2Content="kubeapps"
                title="foo"
              />
            </CardGrid>
          </div>
        </div>
      </div>
    </LoadingWrapper>
  </main>
</div>
`;
